File: common/src/main/kotlin/models/repositories/Repository.kt
/*
 * 🐻‍❄️📦 charted-server: Free, open source, and reliable Helm Chart registry made in Kotlin.
 * Copyright 2022-2023 Noelware, LLC. <team@noelware.org>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.noelware.charted.models.repositories

import com.fasterxml.jackson.annotation.JsonProperty
import io.swagger.v3.oas.annotations.media.Schema
import kotlinx.datetime.LocalDateTime
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import org.noelware.charted.common.types.helm.RepoType
import org.noelware.charted.models.Name
import org.noelware.charted.snowflake.Snowflake

/**
 * Represents a repository that is represented as a singular Helm chart owned by an Organization
 * or User resource.
 */
@Serializable
public data class Repository(
    /**
     * Short description about this user, can be `null` if none was provided.
     */
    @get:Schema(description = "Short description about this user, can be `null` if none was provided.")
    val description: String? = null,

    /**
     * Whether if this repository is deprecated or not
     */
    @get:Schema(description = "Whether if this repository is or not")
    val deprecated: Boolean = false,

    /**
     * Date of when this repository was registered to this instance
     */
    @JsonProperty("created_at")
    @SerialName("created_at")
    @get:Schema(description = "Date of when this repository was registered to this instance")
    val createdAt: LocalDateTime,

    /**
     * Date of when the server has last updated this repository
     */
    @JsonProperty("updated_at")
    @SerialName("updated_at")
    @get:Schema(description = "Date of when the server has last updated this repository")
    val updatedAt: LocalDateTime,

    /**
     * Unique hash to locate a repository's icon, this also includes the extension that this avatar is, i.e, `png`.
     */
    @JsonProperty("icon_hash")
    @SerialName("icon_hash")
    @get:Schema(description = "Unique hash to locate a repository's icon, this also includes the extension that this avatar is, i.e, `png`.")
    val iconHash: String? = null,

    /**
     * Whether if this repository is private or not
     */
    @get:Schema(description = "Whether if this repository is private or not")
    val private: Boolean = false,

    /**
     * Unique identifier that points to a User or Organization resource that owns this repository
     */
    @JsonProperty("owner_id")
    @SerialName("owner_id")
    @get:Schema(description = "Unique identifier that points to a User or Organization resource that owns this repository", implementation = Snowflake::class)
    val ownerID: Long,

    /**
     * Unique [Name] to locate this repository from the API
     */
    @get:Schema(description = "Unique [Name] to locate this repository from the API", implementation = Name::class)
    val name: String,

    /**
     * The chart type that this repository is
     */
    @get:Schema(description = "The chart type that this repository is")
    val type: RepoType,

    /**
     * Unique identifier to locate this repository from the API
     */
    @get:Schema(description = "Unique identifier to locate this repository from the API", implementation = Snowflake::class)
    val id: Long
)

File: crates/types/src/payloads/repository.rs
// 🐻‍❄️📦 charted-server: Free, open source, and reliable Helm Chart registry made in Rust
// Copyright 2022-2024 Noelware, LLC. <team@noelware.org>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

pub mod member;
pub mod release;

use crate::{helm::ChartType, name::Name};
use serde::Deserialize;
use utoipa::ToSchema;

super::create_modifying_payload! {
    Repository;

    /// Request body payload for creating a repository.
    #[derive(Debug, Clone, Deserialize, ToSchema)]
    create {
        /// Short description about this repository.
        #[serde(default)]
        pub description: Option<String>,

        /// Whether if this repository is private.
        #[serde(default)]
        pub private: bool,

        /// The contents of the README that will be displayed on the repository. If you're
        /// using charted's official Helm plugin, new releases can update its README and it'll
        /// be reflected.
        ///
        /// This should be valid Markdown, but XSS cross scripting is impossible as scripts
        /// in codeblocks or via `<script>` won't be executed.
        ///
        /// You can retrieve a repository's README (if it is public or if you have access) with
        /// the [`GET /cdn`](https://charts.noelware.org/docs/server/latest/api/cdn#GET-{...params}) REST
        /// endpoint if the instance has the CDN feature enabled. It'll be under `/repositories/{id}/README.md`.
        #[serde(default)]
        pub readme: Option<String>,

        #[serde(skip)]
        pub creator: i64,

        /// [`Name`] to attach to this repository.
        pub name: Name,

        /// Type of chart this represents. When serializing to valid Helm objects,
        /// `operator` will be replaced with `application`.
        #[serde(default, rename = "type")]
        pub r#type: ChartType,
    }

    /// Request body payload for patching a repository's metadata.
    #[derive(Debug, Clone, Deserialize, ToSchema)]
    patch {
        /// Short description about this repository. If `description` was set to `null`, then
        /// this will not be updated, if `description` was a empty string, the `description`
        /// will be set to a empty string and will present as "*no description for this repository*"
        /// in Hoshi.
        #[serde(default)]
        pub description: Option<String>,

        /// Whether if this repository is private. This cannot be set to the actual value
        /// that it was previously.
        #[serde(default)]
        pub private: Option<bool>,

        /// The contents of the README that will be displayed on the repository. If you're
        /// using charted's official Helm plugin, new releases can update its README and it'll
        /// be reflected.
        ///
        /// This should be valid Markdown, but XSS cross scripting is impossible as scripts
        /// in codeblocks or via `<script>` won't be executed.
        ///
        /// You can retrieve a repository's README (if it is public or if you have access) with
        /// the [`GET /cdn`](https://charts.noelware.org/docs/server/latest/api/cdn#GET-{...params}) REST
        /// endpoint if the instance has the CDN feature enabled. It'll be under `/repositories/{id}/README.md`.
        #[serde(default)]
        pub readme: Option<String>,

        /// [`Name`] to update towards, this will not update if it is
        /// the same.
        #[schema(value_type = Name)]
        pub name: Option<Name>,

        /// Type of chart this represents. When serializing to valid Helm objects,
        /// `operator` will be replaced with `application`.
        #[serde(default, rename = "type")]
        pub r#type: Option<ChartType>,
    }
}
